File: php/src/Advertisement/Domain/Exceptions/AdvertisementNotFoundException.php
<?php
declare(strict_types=1);

namespace Demo\App\Advertisement\Domain\Exceptions;

use Demo\App\Common\Domain\DomainException;

final class AdvertisementNotFoundException extends DomainException
{
    private const string NOT_FOUND_WITH_ID_MESSAGE = 'Advertisement not found with ID: ';
    private const int NOT_FOUND_ERROR_CODE = 404;

    /**
     * @param string $message
     */
    private function __construct(string $message)
    {
        parent::__construct($message, self::NOT_FOUND_ERROR_CODE);
    }

    public static function withId(string $id): self
    {
        return new self(self::NOT_FOUND_WITH_ID_MESSAGE . $id);
    }

    public function message(): string
    {
        return $this->getMessage();
    }
}

File: kotlin/src/main/kotlin/advertisement/domain/exceptions/AdvertisementNotFoundException.kt
package advertisement.domain.exceptions

import common.application.ElementNotFoundException

class AdvertisementNotFoundException private constructor(message: String) : ElementNotFoundException(message) {

    companion object {
        fun withId(id: String): AdvertisementNotFoundException {
            return AdvertisementNotFoundException("Advertisement not found with Id $id")
        }
    }
}
